theano/gpuarray/dnn.py

```diff
@@ -2833,6 +2833,273 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        _, grid = outputs
+        dy = grads[0]
+
+        dimg, dgrid = GpuDnnTransformerGradI()(img, theta, grid, dy, desc)
+        dtheta = GpuDnnTransformerGradT()(dgrid, desc)
+        dgrid_dims = grad_not_implemented(self, grid_dims, 2)
+
+        return [dimg, dtheta, dgrid_dims, DisconnectedType()()]
+
+    def connection_pattern(self, node):
+        # not connected to desc
+        return [[1, 1], [1, 1], [1, 1], [0, 0]]
+
+
+class GpuDnnTransformerGradI(DnnBase):
+    """
+    Gradients of inputs of the spatial transformer
+    """
+    __props__ = ()
+    _cop_num_inputs = 5
+    _cop_num_outputs = 2
+    _f16_ok = True
+
+    def __init__(self, dtype=theano.config.floatX):
+        DnnBase.__init__(self, ["c_code/dnn_sptf_gi.c"], "APPLY_SPECIFIC(dnn_sptf_gi)")
+
+    def make_node(self, img, theta, grid, dy, desc):
+        context_name = infer_context_name(img, theta, grid, dy, desc)
+
+        if (not isinstance(desc.type, CDataType) or
+                desc.type.ctype != 'cudnnSpatialTransformerDescriptor_t'):
+            raise ValueError('desc must be cudnnSpatialTransformerDescriptor_t')
+
+        img = as_gpuarray_variable(gpu_contiguous(img), context_name)
+        if img.ndim != 4:
+            raise TypeError('img must have 4 dimensions.')
+
+        theta = as_gpuarray_variable(gpu_contiguous(theta), context_name)
+        if theta.ndim != 3:
+            raise TypeError('theta must have 3 dimensions')
+
+        grid = as_gpuarray_variable(gpu_contiguous(grid), context_name)
+        if img.ndim != grid.ndim:
+            raise TypeError('grid should have the same number of dimensions as img')
+
+        dy = as_gpuarray_variable(dy, context_name)
+        if dy.ndim != 4:
+            raise TypeError('dy must have 4 dimensions.')
+
+        dimg = img.type()
+        dgrid = grid.type()
+
+        inputs = [img, theta, grid, dy, desc]
+        outputs = [dimg, dgrid]
+
+        return Apply(self, inputs, outputs)
+
+    def L_op(self, inputs, outputs, grads):
+        img, theta, grid, dy, desc = inputs
+        dimg_out, dgrid = outputs
+        grad_cost = grads[0]
+
+        dimg = dimg_out * grad_cost
+        dtheta = GpuDnnTransformerGradT()(dgrid, desc)
+        d_dy = grad_not_implemented(self, dy, 3)
+
+        return [dimg, dtheta, dgrid, d_dy, DisconnectedType()()]
+
+    def connection_pattern(self, node):
+        # not connected to desc
+        return [[1, 1], [1, 1], [1, 1], [1, 1], [0, 0]]
+
+
+class GpuDnnTransformerGradT(DnnBase):
+    """
+    Gradients of the affine transformation generated by the localisation network
+    """
+    __props__ = ()
+    _cop_num_inputs = 2
+    _cop_num_outputs = 1
+    _f16_ok = True
+
+    def __init__(self):
+        DnnBase.__init__(self, ["c_code/dnn_sptf_gt.c"], "APPLY_SPECIFIC(dnn_sptf_gt)")
+
+    def make_node(self, dgrid, desc):
+        context_name = infer_context_name(dgrid)
+        dtheta = GpuArrayType(dtype=dgrid.dtype,
```

Comment:
Before this line:
- apply `as_gpuarray_variable` to dgrid.
- check dgrid dtype (in float16, float32 or float64)
- check dgrid ndim (expected 4).